### ISSUE TITLE
Port RPC functionality from Bitcoin

### ISSUE THREAD

**Issue by denravonska** [gridcoin-community/Gridcoin-Research#1142]
Copy Bitcoin code
==============
Copy the relevant Bitcoin RPC code into `rpc/`. This will cascade into dependencies so if the job is not feasible to do we will notice it here.

Gridcoin RPC isolation
=================
With the new RPC engine in place we can migrate our Gridcoin specific calls from the old RPC files under `src/rpc*` to files following the Bitcoin naming convention under `src/gridcoin/rpc/`. We can and should be liberal with axing calls which are unused or obsolete calls which can be done by external scripts using existing calls.

RPC table injection
===============
Bitcoin calls `RegisterAllCoreRPCCommands` during initialization to register the RPC functions. At the end of this we should include a call to `RegisterGridcoinRPCCommands` which is implemented in `gridcoin/rpc/register.h` and follows the same pattern but only injects Gridcoin specific commands.

**Comment by RoboticMind** [comment]
I think at least part 1 and 2 would be covered by https://github.com/gridcoin-community/Gridcoin-Research/pull/1705 whenever that gets merged. Not sure about the RPC table injection thing

**Comment by PrestackI** [comment]
I had AI review the BTC Repo RPC Functionality and then compared it to GRC repo for RPC functionality as of Today (December 27th 2025). 
**The report is below.**

I had it prioritize what it thought was most important. 
I have no idea how to evaluate the prioritization so I am looking for feedback on what the community thinks is important. 

I do not intend to take action on any of these at this moment - this report is for discussion purposes.


# Gridcoin RPC Porting Analysis Report
## Bitcoin Core RPC Functionality Review

**Report Date**: December 27, 2025  
**Gridcoin Version**: 5.4.9.0 (development)  
**Bitcoin Reference**: Master branch (December 2025)  
**Analysis Scope**: Issue #1142 - Port RPC functionality from Bitcoin

---

## Executive Summary

This report analyzes the current state of RPC command implementation in Gridcoin compared to Bitcoin Core, identifies gaps, and provides a prioritized roadmap for porting missing functionality.

### Key Findings

**Current Status:**
- **Total Gridcoin RPC Commands**: ~140 commands across 5 categories
- **Organization**: Partially modularized (10 files in `src/rpc/`)
- **Framework**: Uses older actor/function pointer pattern
- **Previous Work**: Issue #1705 and PR #1995 created initial `src/rpc/` directory structure (v5.2.0.0)

**Major Gaps Identified:**
1. ❌ Modern `RPCHelpMan` self-documenting framework
2. ❌ Dedicated mempool management commands
3. ❌ Fee estimation commands (Gridcoin uses flat fees)
4. ❌ PSBT (Partially Signed Bitcoin Transaction) support
5. ❌ Descriptor wallet functionality
6. ❌ Separate wallet RPC directory structure
7. ❌ Node management commands
8. ❌ Transaction proof commands

**Recommendations:**
- **Priority 1**: Adopt RPCHelpMan framework for better documentation
- **Priority 2**: Add missing mempool and network diagnostic commands
- **Priority 3**: Enhance wallet commands with modern Bitcoin features (where applicable)
- **Priority 4**: Consider PSBT support for multi-signature workflows

---

## Section 1: Current Gridcoin RPC Structure

### 1.1 File Organization

```
src/rpc/
├── blockchain.cpp       # Block and chain queries
├── client.cpp          # Client-side utilities
├── dataacq.cpp         # Data acquisition (Gridcoin-specific)
├── mining.cpp          # Staking/mining operations
├── misc.cpp            # Miscellaneous commands
├── net.cpp             # Network operations
├── protocol.cpp        # Protocol definitions
├── rawtransaction.cpp  # Raw transaction handling
├── server.cpp          # RPC server and command registration
└── voting.cpp          # Voting system (Gridcoin-specific)
```

**Note**: Unlike Bitcoin Core, Gridcoin does not have a separate `src/wallet/rpc/` directory.

### 1.2 Command Categories

Gridcoin organizes commands into 5 categories:

1. **cat_wallet** - Wallet operations (49 commands)
2. **cat_staking** - Research/CPID/Beacon/Staking (15 commands)
3. **cat_developer** - Development and debugging (39 commands)
4. **cat_network** - Network and blockchain queries (30 commands)
5. **cat_voting** - Poll and voting system (7 commands)

---

## Section 2: Detailed Command Inventory

### 2.1 Wallet Commands (49 commands)

#### Address Management
- `addmultisigaddress` - Create multisig address
- `addredeemscript` - Add redeemscript to wallet
- `getaccount` - Get account for address (deprecated)
- `getaccountaddress` - Get address for account (deprecated)
- `getaddressesbyaccount` - List addresses for account (deprecated)
- `getnewaddress` - Generate new address
- `getnewpubkey` - Generate new public key
- `setaccount` - Set account for address (deprecated)
- `validateaddress` - Validate address
- `validatepubkey` - Validate public key

#### Balance & Wallet Info
- `getbalance` - Get wallet balance
- `getbalancedetail` - Get detailed balance breakdown
- `getreceivedbyaccount` - Amount received by account (deprecated)
- `getreceivedbyaddress` - Amount received by address
- `getunconfirmedbalance` - Get unconfirmed balance
- `getwalletinfo` - Get wallet information
- `listaccounts` - List accounts (deprecated)
- `listaddressgroupings` - List address groupings
- `listreceivedbyaccount` - List received by account (deprecated)
- `listreceivedbyaddress` - List received by address
- `reservebalance` - Reserve balance for non-staking

#### Transaction Management
- `gettransaction` - Get transaction details
- `getrawtransaction` - Get raw transaction
- `getrawwallettransaction` - Get raw wallet transaction
- `listsinceblock` - List transactions since block
- `liststakes` - List staking transactions
- `listtransactions` - List transactions
- `listunspent` - List unspent outputs
- `resendtx` - Resend transaction

#### Transaction Creation & Signing
- `createrawtransaction` - Create raw transaction
- `decoderawtransaction` - Decode raw transaction
- `decodescript` - Decode script
- `sendrawtransaction` - Broadcast raw transaction
- `signmessage` - Sign message with private key
- `signrawtransaction` - Sign raw transaction
- `verifymessage` - Verify signed message

#### Sending
- `burn` - Burn coins
- `rainbymagnitude` - Send by magnitude (Gridcoin-specific)
- `sendfrom` - Send from account
- `sendmany` - Send to multiple addresses
- `sendtoaddress` - Send to address

#### Key Management
- `dumpprivkey` - Export private key
- `dumpwallet` - Export wallet
- `encryptwallet` - Encrypt wallet
- `importprivkey` - Import private key
- `importwallet` - Import wallet
- `keypoolrefill` - Refill keypool
- `makekeypair` - Generate key pair
- `sethdseed` - Set HD seed
- `walletlock` - Lock wallet
- `walletpassphrase` - Unlock wallet temporarily
- `walletpassphrasechange` - Change wallet passphrase

#### Wallet Maintenance
- `backupwallet` - Backup wallet
- `checkwallet` - Check wallet integrity
- `maintainbackups` - Manage automatic backups
- `repairwallet` - Repair wallet
- `upgradewallet` - Upgrade wallet format
- `walletdiagnose` - Diagnose wallet issues

#### Utility
- `consolidatemsunspent` - Consolidate multisig unspent
- `consolidateunspent` - Consolidate unspent outputs
- `move` - Move between accounts (deprecated)
- `scanforunspent` - Scan for unspent outputs
- `settxfee` - Set transaction fee

### 2.2 Staking Commands (15 commands)

These are Gridcoin-specific commands for the proof-of-stake and research reward system:

#### Beacon Management
- `advertisebeacon` - Advertise research beacon
- `beaconconvergence` - Check beacon convergence
- `beaconreport` - Generate beacon report
- `beaconstatus` - Check beacon status
- `pendingbeaconreport` - Report pending beacons
- `revokebeacon` - Revoke beacon

#### Research & Magnitude
- `explainmagnitude` - Explain magnitude calculation
- `lifetime` - Get lifetime research statistics
- `magnitude` - Get CPID magnitude
- `resetcpids` - Reset CPID detection

#### Staking Info
- `getlaststake` - Get information about last stake
- `getstakinginfo` - Get staking information
- `getmininginfo` - Alias for getstakinginfo

#### Manual Reward Claims (MRC)
- `createmrcrequest` - Create manual reward claim
- `getmrcinfo` - Get MRC information

#### Superblock
- `superblockage` - Get superblock age
- `superblocks` - List superblocks

### 2.3 Developer Commands (39 commands)

#### Debugging & Diagnostics
- `addkey` - Add key to registry
- `auditsnapshotaccrual` - Audit snapshot accrual
- `auditsnapshotaccruals` - Audit multiple snapshot accruals
- `beaconaudit` - Audit beacon system
- `changesettings` - Change settings dynamically
- `debug` - Control debug flags
- `inspectaccrualsnapshot` - Inspect accrual snapshot
- `listalerts` - List network alerts
- `logging` - Control logging categories
- `parseaccrualsnapshotfile` - Parse accrual snapshot file
- `versionreport` - Generate version report
- `walletdiagnose` - Diagnose wallet issues

#### Contract & Registry Management
- `currentcontractaverage` - Get contract average
- `dumpcontracts` - Dump contracts
- `listprotocolentries` - List protocol entries
- `listsettings` - List settings

#### Project & Research
- `getautogreylist` - Get auto-greylist status
- `listprojects` - List whitelisted projects
- `listresearcheraccounts` - List researcher accounts
- `projects` - Get project information

#### Scraper (Gridcoin-specific)
- `archivelog` - Archive log files
- `convergencereport` - Get convergence report
- `deletecscrapermanifest` - Delete scraper manifest
- `getmpart` - Get manifest part
- `listmanifests` - List scraper manifests
- `listscrapers` - List active scrapers
- `savescraperfilemanifest` - Save scraper manifest
- `scraperreport` - Get scraper report
- `sendscraperfilemanifest` - Send scraper manifest
- `testnewsb` - Test new superblock

#### Sidestaking
- `listmandatorysidestakes` - List mandatory sidestakes

#### Block Statistics & Analysis
- `exportstats1` - Export statistics
- `getblockstats` - Get block statistics
- `getrecentblocks` - Get recent blocks
- `parselegacysb` - Parse legacy superblock
- `superblockaverage` - Get superblock average

#### Network & Data
- `network` - Network statistics
- `readdata` - Read data from registry
- `writedata` - Write data to registry
- `reorganize` - Reorganize blockchain

#### Alert Management
- `sendalert` - Send network alert
- `sendalert2` - Send network alert (v2)
- `sendblock` - Send block to peer

### 2.4 Network Commands (30 commands)

#### Peer Management
- `addnode` - Add/remove/connect to node
- `clearbanned` - Clear banned nodes
- `getaddednodeinfo` - Get added node information
- `getnodeaddresses` - Get node addresses
- `getconnectioncount` - Get connection count
- `getpeerinfo` - Get peer information
- `listbanned` - List banned nodes
- `ping` - Ping connected nodes
- `setban` - Ban/unban IP address

#### Network Information
- `getnetworkinfo` - Get network information
- `getnettotals` - Get network traffic totals
- `currenttime` - Get current time
- `networktime` - Get network-adjusted time

#### Blockchain Queries
- `askforoutstandingblocks` - Request outstanding blocks
- `getbestblockhash` - Get best block hash
- `getblock` - Get block data
- `getblockbynumber` - Get block by number
- `getblockbymintime` - Get block by minimum time
- `getblockcount` - Get block count
- `getblockhash` - Get block hash by height
- `getblocksbatch` - Get multiple blocks
- `getcheckpoint` - Get checkpoint information
- `showblock` - Show block details

#### Blockchain Information
- `getblockchaininfo` - Get blockchain state
- `getburnreport` - Get burn report
- `getdifficulty` - Get proof-of-stake difficulty
- `getinfo` - Get general information
- `getrawmempool` - Get raw mempool contents

#### Control
- `stop` - Stop Gridcoin server

### 2.5 Voting Commands (7 commands)

Gridcoin's decentralized governance system:

- `addpoll` - Create new poll
- `getpollresults` - Get poll results
- `getvotingclaim` - Get voting claim
- `listpolls` - List all polls
- `testpollnotification` - Test poll notification
- `vote` - Vote in poll (deprecated)
- `votebyid` - Vote in poll by ID
- `votedetails` - Get vote details

---

## Section 3: Bitcoin Core RPC Categories - Comparison

### 3.1 Blockchain Commands

**Bitcoin Core** (`src/rpc/blockchain.cpp`):
- 20+ commands for blockchain queries and management
- Advanced UTXO set operations
- Blockchain verification tools
- Snapshot creation/loading

**Gridcoin Status**:
✅ **Has (similar functionality)**:
- `getbestblockhash`, `getblock`, `getblockcount`, `getblockhash`, `getblockchaininfo`
- `getdifficulty`, `getcheckpoint`

❌ **Missing**:
- `getchaintips` - Information about all chain tips
- `gettxout` - Details about unspent transaction output
- `gettxoutsetinfo` - Statistics about UTXO set
- `pruneblockchain` - Prune blockchain to specific height
- `verifychain` - Verify blockchain database
- `scantxoutset` - Scan UTXO set for descriptors
- `scanblocks` - Scan blocks for relevant transactions
- `dumptxoutset` - Dump UTXO set to file
- `loadtxoutset` - Load UTXO set from file

**Assessment**: Moderate gap. Missing advanced UTXO set tools.

### 3.2 Mining/Staking Commands

**Bitcoin Core** (`src/rpc/mining.cpp`):
- Mining operations and block generation
- Transaction prioritization
- Block template creation (BIP 22/23)

**Gridcoin Status**:
✅ **Has (adapted for PoS)**:
- `getstakinginfo` (equivalent to getmininginfo)
- Staking-specific commands (beacons, magnitude, etc.)

⚠️ **Partially implemented**:
- Transaction prioritization not exposed via RPC

❌ **Missing**:
- `getnetworkhashps` - N/A for PoS (not applicable)
- `prioritisetransaction` - Modify transaction priority
- `getprioritisedtransactions` - List prioritised transactions  
- `getblocktemplate` - Get block template (used by pools)
- `submitblock` - Submit solved block
- `submitheader` - Submit block header
- Mining generation commands (regtest): `generatetoaddress`, `generatetodescriptor`, `generateblock`

**Assessment**: Gridcoin has adapted these for PoS. Some mining-specific features not needed, but transaction prioritization could be useful.

### 3.3 Mempool Commands

**Bitcoin Core** (`src/rpc/mempool.cpp`):
- Comprehensive mempool management
- Transaction package submission
- Ancestor/descendant analysis

**Gridcoin Status**:
✅ **Has**:
- `getrawmempool` - Basic mempool listing
- `sendrawtransaction` - Submit transaction

❌ **Missing**:
- `getmempoolinfo` - Mempool statistics
- `getmempoolentry` - Specific mempool entry details
- `getmempoolancestors` - Get in-mempool ancestors
- `getmempooldescendants` - Get in-mempool descendants
- `getmempoolcluster` - Get transaction cluster info
- `savemempool` - Save mempool to disk
- `importmempool` - Load mempool from file
- `testmempoolaccept` - Test transaction acceptance
- `submitpackage` - Submit package of transactions
- `getorphantxs` - Get orphan transactions

**Assessment**: **MAJOR GAP**. Gridcoin has minimal mempool RPC functionality.

### 3.4 Network Commands

**Bitcoin Core** (`src/rpc/net.cpp`):
- Extensive peer management
- Address manager diagnostics
- Network testing tools

**Gridcoin Status**:
✅ **Has**:
- `addnode`, `getaddednodeinfo`, `getconnectioncount`, `getpeerinfo`
- `getnetworkinfo`, `getnettotals`, `getnodeaddresses`
- `setban`, `listbanned`, `clearbanned`
- `ping`

❌ **Missing**:
- `addconnection` - Add test connection
- `disconnectnode` - Disconnect from node
- `setnetworkactive` - Enable/disable network activity
- `addpeeraddress` - Add peer address to address manager
- `sendmsgtopeer` - Send message to peer (test only)
- `getaddrmaninfo` - Get address manager info
- `getrawaddrman` - Get raw address manager data

**Assessment**: Moderate gap. Missing diagnostic and testing tools.

### 3.5 Node Management Commands

**Bitcoin Core** (`src/rpc/node.cpp`):
- Node management and system information
- Memory diagnostics
- Logging control

**Gridcoin Status**:
✅ **Has**:
- `stop` - Stop server
- `logging` - Control logging (Gridcoin has this)

❌ **Missing**:
- `getmemoryinfo` - Memory usage information
- `setmocktime` - Set mock time (regtest)
- `mockscheduler` - Mock scheduler (regtest)
- `echo` / `echojson` - Echo arguments (test)
- `getindexinfo` - Get index information
- `uptime` - Get server uptime
- `getrpcinfo` - Get RPC execution statistics

**Assessment**: Minor gap. Missing diagnostic utilities.

### 3.6 Raw Transaction Commands

**Bitcoin Core** (`src/rpc/rawtransaction.cpp`):
- Raw transaction creation and manipulation
- PSBT (Partially Signed Bitcoin Transactions) support

**Gridcoin Status**:
✅ **Has**:
- `createrawtransaction`, `decoderawtransaction`, `getrawtransaction`
- `signrawtransaction` (with keys)
- `sendrawtransaction`
- `decodescript`

❌ **Missing - PSBT Support**:
- `createpsbt` - Create PSBT
- `converttopsbt` - Convert transaction to PSBT
- `utxoupdatepsbt` - Update PSBT with UTXO data
- `combinepsbt` - Combine PSBTs
- `finalizepsbt` - Finalize PSBT
- `decodepsbt` - Decode PSBT
- `analyzepsbt` - Analyze PSBT
- `joinpsbts` - Join multiple PSBTs
- `descriptorprocesspsbt` - Process PSBT with descriptor

❌ **Missing - Other**:
- `combinerawtransaction` - Combine partial transactions
- `signrawtransactionwithkey` - Sign with specific keys (has older version)

**Assessment**: **MAJOR GAP**. No PSBT support, which is important for multisig workflows.

### 3.7 Output Script / Address Commands

**Bitcoin Core** (`src/rpc/output_script.cpp`):
- Address and script validation
- Descriptor support

**Gridcoin Status**:
✅ **Has**:
- `validateaddress` - Validate address
- `validatepubkey` - Validate pubkey

❌ **Missing**:
- `createmultisig` - Create multisig address (Gridcoin has `addmultisigaddress` but not `createmultisig`)
- `getdescriptorinfo` - Get descriptor information
- `deriveaddresses` - Derive addresses from descriptor

**Assessment**: Minor gap. Missing descriptor wallet features.

### 3.8 Fee Estimation Commands

**Bitcoin Core** (`src/rpc/fees.cpp`):
- Dynamic fee estimation

**Gridcoin Status**:
❌ **Missing (Not Applicable)**:
- `estimatesmartfee` - Estimate fee for confirmation target
- `estimaterawfee` - Detailed fee estimation

**Assessment**: Not applicable - Gridcoin uses flat fee structure.

### 3.9 Message Signing Commands

**Bitcoin Core** (`src/rpc/signmessage.cpp`):
- Message signing and verification

**Gridcoin Status**:
✅ **Has**:
- `signmessage` - Sign message
- `verifymessage` - Verify message
- `signmessagewithprivkey` - Sign with specific private key

**Assessment**: Complete parity.

### 3.10 Transaction Proof Commands

**Bitcoin Core** (`src/rpc/txoutproof.cpp`):
- Merkle proof creation and verification

**Gridcoin Status**:
❌ **Missing**:
- `gettxoutproof` - Get merkle proof for transaction
- `verifytxoutproof` - Verify merkle proof

**Assessment**: Missing, but low priority feature.

### 3.11 External Signer Commands

**Bitcoin Core** (`src/rpc/external_signer.cpp`):
- Hardware wallet integration

**Gridcoin Status**:
❌ **Missing**:
- `enumeratesigners` - List connected hardware wallets

**Assessment**: Missing, low priority for current development.

### 3.12 Wallet Commands

**Bitcoin Core** (`src/wallet/rpc/` - separate directory):
- Organized into: wallet.cpp, addresses.cpp, backup.cpp, coins.cpp, encrypt.cpp, signmessage.cpp, spend.cpp, transactions.cpp

**Gridcoin Status**:
✅ **Has most core wallet functionality**
- Address management, balance queries, transaction management
- Backup and restore, encryption
- Sending and receiving

⚠️ **Different implementation**:
- No separate `src/wallet/rpc/` directory
- All wallet commands in main `src/rpc/` files

❌ **Missing modern features**:
- Descriptor wallet support
- `getbalances` - Get all balances (plural)
- `gettransaction` doesn't have all Bitcoin fields
- Watch-only wallet features limited
- No `psbtbumpfee` or other PSBT wallet integration

**Assessment**: Good coverage but lacks modern descriptor wallet features.

---

## Section 4: Gap Analysis Summary

### 4.1 Framework and Architecture Gaps

#### 4.1.1 RPCHelpMan Framework
**Status**: ❌ **NOT IMPLEMENTED**

**What It Is**:
- Modern self-documenting RPC command framework from Bitcoin Core
- Provides automatic help text generation
- Type-safe argument extraction
- Compile-time argument validation
- Structured documentation (args, results, examples)

**Current Gridcoin Approach**:
- Uses older `actor` function pointer pattern
- Manual help text in each function
- Manual argument parsing with `params[n]` indexing
- No structured documentation format

**Impact**: 
- **High** - Harder to maintain, document, and extend RPC commands
- Makes it more difficult to keep help text accurate and comprehensive
- More error-prone when adding new commands

**Difficulty to Port**: **HARD**
- Requires significant refactoring of all RPC commands
- Need to port helper classes: `RPCHelpMan`, `RPCArg`, `RPCResult`, `RPCExamples`
- Must update command registration system
- Each command needs individual conversion

**Recommendation**: High priority for long-term maintainability

---

#### 4.1.2 Modular File Organization
**Status**: ⚠️ **PARTIALLY IMPLEMENTED**

**Bitcoin Structure**:
```
src/rpc/                    # Core RPC (15+ files)
src/wallet/rpc/             # Wallet RPC (8+ files)
```

**Gridcoin Structure**:
```
src/rpc/                    # All RPC (10 files)
```

**Missing**:
- Separate `src/wallet/rpc/` directory
- Dedicated files: `mempool.cpp`, `fees.cpp`, `node.cpp`, `output_script.cpp`, `txoutproof.cpp`, `external_signer.cpp`

**Impact**: 
- **Medium** - Code organization less clear
- Harder to navigate for contributors

**Difficulty to Port**: **EASY to MEDIUM**
- Mostly organizational refactoring
- Move wallet commands to separate directory
- Create new category files

**Recommendation**: Medium priority, good cleanup task

---

### 4.2 Command Category Gaps

#### 4.2.1 Mempool Commands
**Status**: ❌ **MAJOR GAP**

**Missing Commands** (10 commands):
- `getmempoolinfo` - Mempool statistics
- `getmempoolentry` - Entry details
- `getmempoolancestors` - Ancestor analysis
- `getmempooldescendants` - Descendant analysis  
- `getmempoolcluster` - Cluster information
- `savemempool` / `importmempool` - Persistence
- `testmempoolaccept` - Transaction testing
- `submitpackage` - Package submission
- `getorphantxs` - Orphan transactions

**Use Cases**:
- Debugging stuck transactions
- Fee estimation (though Gridcoin uses flat fees)
- Transaction relationship analysis
- Pool operators and power users

**Impact**: 
- **HIGH** - Important diagnostic tools missing
- Harder to debug mempool issues
- Less visibility into pending transactions

**Difficulty to Port**: **MEDIUM to HARD**
- Core mempool functionality exists in Gridcoin
- Need to expose via RPC interface
- Some features (package relay) may need additional work

**Recommendation**: **HIGH PRIORITY** - Add at least basic mempool diagnostic commands

---

#### 4.2.2 PSBT (Partially Signed Bitcoin Transactions)
**Status**: ❌ **MISSING**

**Missing Commands** (9 commands):
- `createpsbt`, `converttopsbt`, `utxoupdatepsbt`
- `combinepsbt`, `finalizepsbt`
- `decodepsbt`, `analyzepsbt`
- `joinpsbts`, `descriptorprocesspsbt`

**What PSBT Provides**:
- Standard format for unsigned/partially signed transactions
- Essential for multisig workflows
- Hardware wallet integration
- Air-gapped signing
- Multi-party transaction construction

**Use Cases**:
- Multisig wallets (important for Gridcoin Foundation, pools)
- Cold storage signing
- Collaborative transaction creation
- Future hardware wallet support

**Impact**: 
- **HIGH** - Missing modern multisig workflow support
- Makes complex multisig operations harder
- Limits hardware wallet integration options

**Difficulty to Port**: **VERY HARD**
- Requires implementing PSBT library/format (BIP 174)
- Significant code addition (~2000-3000 lines)
- Need to understand PSBT format thoroughly
- Integration with existing wallet code

**Recommendation**: **HIGH PRIORITY** for multisig users, but significant effort required

---

#### 4.2.3 Blockchain UTXO Analysis Commands
**Status**: ⚠️ **PARTIAL GAP**

**Missing Commands** (9 commands):
- `getchaintips` - Chain tip information
- `gettxout` - UTXO details
- `gettxoutsetinfo` - UTXO set statistics
- `pruneblockchain` - Blockchain pruning
- `verifychain` - Database verification
- `scantxoutset` - UTXO set scanning
- `scanblocks` - Block scanning
- `dumptxoutset` / `loadtxoutset` - UTXO snapshots

**Use Cases**:
- Debugging blockchain issues
- Performance analysis
- Disk space management (pruning)
- Fast synchronization (UTXO snapshots)

**Impact**: 
- **MEDIUM** - Useful diagnostic and optimization tools missing
- Pruning would help with disk space concerns
- UTXO snapshots could speed up sync

**Difficulty to Port**: 
- **EASY to MEDIUM**: `getchaintips`, `gettxout`, `verifychain`
- **HARD**: `pruneblockchain`, `scantxoutset`, `dumptxoutset`/`loadtxoutset`

**Recommendation**: MEDIUM PRIORITY - Add easier diagnostic commands first

---

#### 4.2.4 Network Diagnostic Commands  
**Status**: ⚠️ **PARTIAL GAP**

**Missing Commands** (7 commands):
- `addconnection` - Test connections
- `disconnectnode` - Force disconnect
- `setnetworkactive` - Enable/disable networking
- `addpeeraddress` - Address manager manipulation
- `sendmsgtopeer` - Send test messages
- `getaddrmaninfo` - Address manager diagnostics
- `getrawaddrman` - Raw address manager data

**Use Cases**:
- Network debugging
- Testing network code
- Diagnosing connection issues
- Developer tools

**Impact**: 
- **LOW to MEDIUM** - Mostly development/debugging tools
- Would help developers and power users troubleshoot

**Difficulty to Port**: **EASY to MEDIUM**
- Most functionality already exists
- Just need RPC exposure
- Some are test-only commands

**Recommendation**: LOW to MEDIUM PRIORITY - Nice to have for debugging

---

#### 4.2.5 Node Management Commands
**Status**: ⚠️ **PARTIAL GAP**

**Missing Commands** (7 commands):
- `getmemoryinfo` - Memory usage stats
- `setmocktime` / `mockscheduler` - Testing tools (regtest)
- `echo` / `echojson` - Echo test
- `getindexinfo` - Index information
- `uptime` - Server uptime
- `getrpcinfo` - RPC statistics

**Use Cases**:
- Performance monitoring (`getmemoryinfo`, `getrpcinfo`)
- System information (`uptime`)
- Development/testing (mock commands, echo)
- Index status monitoring

**Impact**: 
- **LOW to MEDIUM** - Useful diagnostic information
- `getmemoryinfo` and `getrpcinfo` would be helpful for debugging

**Difficulty to Port**: **EASY**
- Straightforward information reporting
- Minimal new functionality required

**Recommendation**: LOW PRIORITY - Easy wins for additional diagnostics

---

#### 4.2.6 Transaction Proof Commands
**Status**: ❌ **MISSING**

**Missing Commands** (2 commands):
- `gettxoutproof` - Create merkle proof
- `verifytxoutproof` - Verify merkle proof

**Use Cases**:
- SPV (Simplified Payment Verification) clients
- Proving transaction inclusion without full blockchain
- Cross-chain verification

**Impact**: 
- **LOW** - Niche use case
- Not critical for typical Gridcoin operations

**Difficulty to Port**: **EASY to MEDIUM**
- Merkle tree logic already exists
- Just need to expose via RPC

**Recommendation**: LOW PRIORITY - Nice to have but not essential

---

#### 4.2.7 Descriptor Wallet Support
**Status**: ❌ **MISSING**

**Missing Commands** (3 commands):
- `getdescriptorinfo` - Parse descriptor
- `deriveaddresses` - Derive from descriptor
- `createmultisig` - Create multisig (Gridcoin has `addmultisigaddress`)

**What Descriptors Are**:
- Modern way to describe scripts/addresses
- Supports complex derivation paths
- Better for HD wallets
- More flexible than traditional wallet

**Use Cases**:
- Modern wallet management
- Watch-only wallets
- Complex script templates
- HD wallet operations

**Impact**: 
- **MEDIUM** - Modern wallet feature
- Would improve wallet flexibility
- Not critical for basic operations

**Difficulty to Port**: **HARD**
- Requires implementing descriptor parsing library
- Significant wallet code changes
- Depends on understanding Bitcoin's descriptor format

**Recommendation**: MEDIUM PRIORITY - Future enhancement for wallet modernization

---

#### 4.2.8 Transaction Prioritization
**Status**: ⚠️ **PARTIAL** (exists in code, not exposed via RPC)

**Missing Commands** (2 commands):
- `prioritisetransaction` - Adjust transaction priority
- `getprioritisedtransactions` - List prioritized transactions

**Use Cases**:
- Miners/stakers prioritizing specific transactions
- Emergency transaction processing
- Pool operators

**Impact**: 
- **LOW to MEDIUM** - Useful for pool operators
- Less critical with Gridcoin's flat fee structure
- Could still be useful for network management

**Difficulty to Port**: **EASY**
- Gridcoin has priority system in code
- Just need RPC exposure

**Recommendation**: LOW to MEDIUM PRIORITY - Easy addition if there's demand

---

### 4.3 Feature Comparison Matrix

| Feature Category | Bitcoin Core | Gridcoin | Gap Severity | Port Difficulty |
|------------------|-------------|----------|--------------|-----------------|
| **Framework** |
| RPCHelpMan | ✅ Full | ❌ None | HIGH | HARD |
| Modular Organization | ✅ Full | ⚠️ Partial | MEDIUM | EASY-MED |
| **Commands** |
| Blockchain Basic | ✅ Full | ✅ Full | NONE | - |
| Blockchain Advanced | ✅ Full | ⚠️ Partial | MEDIUM | EASY-HARD |
| Mempool | ✅ Full | ❌ Minimal | **HIGH** | MEDIUM-HARD |
| Network Basic | ✅ Full | ✅ Full | NONE | - |
| Network Advanced | ✅ Full | ⚠️ Partial | MEDIUM | EASY-MED |
| Node Management | ✅ Full | ⚠️ Partial | LOW-MED | EASY |
| Raw Transactions | ✅ Full | ✅ Good | LOW | - |
| **PSBT** | ✅ Full | ❌ None | **HIGH** | **VERY HARD** |
| Wallet Basic | ✅ Full | ✅ Full | NONE | - |
| **Descriptor Wallet** | ✅ Full | ❌ None | MEDIUM | HARD |
| Transaction Proofs | ✅ Full | ❌ None | LOW | EASY-MED |
| Message Signing | ✅ Full | ✅ Full | NONE | - |
| Fee Estimation | ✅ Full | N/A | N/A | N/A |
| External Signer | ✅ Full | ❌ None | LOW | MEDIUM |

**Legend**:
- ✅ Full: Complete implementation
- ⚠️ Partial: Some features implemented
- ❌ None: Not implemented
- N/A: Not applicable to Gridcoin

---

### 4.4 Gridcoin-Specific Additions Not in Bitcoin

Gridcoin has implemented many RPC commands that don't exist in Bitcoin Core:

**Staking & Research** (15 commands):
- Beacon management (6 commands)
- Magnitude and research accounting (4 commands)
- MRC (Manual Reward Claims) (2 commands)
- Superblock queries (3 commands)

**Voting System** (7 commands):
- Poll creation and voting
- Result tallying and reporting

**Scraper System** (10 commands):
- Manifest management
- Convergence reporting
- Statistics collection

**Developer Tools** (additional):
- Accrual auditing and inspection
- Contract dumping and analysis
- Protocol entry management
- Greylist management

**Assessment**: Gridcoin has a rich set of commands for its unique PoS/research system that Bitcoin doesn't need. The gap is primarily in general-purpose Bitcoin RPC features, not in Gridcoin-specific functionality.

---

## Section 5: Priority-Ranked Implementation Roadmap

### 5.1 Priority Classification

Commands and features are ranked by:
- **Impact**: How important is this feature for users/developers?
- **Difficulty**: How hard is it to implement?
- **Dependencies**: What other work is required first?
- **Community Need**: Is there user demand for this?

**Priority Levels**:
- **P0 - Critical**: Essential infrastructure improvements
- **P1 - High**: Important functionality gaps
- **P2 - Medium**: Useful enhancements
- **P3 - Low**: Nice-to-have features

---

### 5.2 P0 - Critical Priority (Foundation Work)

These items improve the overall RPC infrastructure and make future work easier.

#### P0.1: Adopt RPCHelpMan Framework
**Difficulty**: ⭐⭐⭐⭐ (HARD)  
**Estimated Effort**: 3-6 months (incremental)  
**Impact**: Very High - Improves all RPC commands

**Description**:
Port Bitcoin's `RPCHelpMan` self-documenting framework to replace the current actor pattern.

**Benefits**:
- Automatic help text generation
- Type-safe argument handling
- Better documentation consistency
- Easier to add new commands
- Compile-time validation

**Implementation Approach**:
1. Port core classes: `RPCHelpMan`, `RPCArg`, `RPCResult`, `RPCExamples`
2. Update command registration in `server.cpp`
3. Convert commands incrementally (start with simple ones)
4. Priority order: Utility commands → Blockchain → Network → Wallet

**Files to Create/Modify**:
- `src/rpc/util.h/cpp` - Port RPCHelpMan framework
- `src/rpc/server.h/cpp` - Update registration
- All command files - Convert individual commands

**Dependencies**: None - can start immediately

**Reference**:
- Bitcoin: `src/rpc/util.h/cpp`
- Examples in all Bitcoin RPC files

---

### 5.3 P1 - High Priority (Important Functionality)

#### P1.1: Mempool Diagnostic Commands
**Difficulty**: ⭐⭐⭐ (MEDIUM-HARD)  
**Estimated Effort**: 2-4 weeks  
**Impact**: High - Important for debugging

**Commands to Add** (in priority order):
1. ✅ `getmempoolinfo` - Basic statistics (EASY, 1-2 days)
2. ✅ `getmempoolentry` - Individual entry details (EASY, 1-2 days)
3. ✅ `testmempoolaccept` - Test transaction (MEDIUM, 3-5 days)
4. ⚠️ `getmempoolancestors` - Ancestor analysis (MEDIUM, 3-5 days)
5. ⚠️ `getmempooldescendants` - Descendant analysis (MEDIUM, 3-5 days)
6. ⚠️ `savemempool` / `importmempool` - Persistence (MEDIUM, 5-7 days)

**Implementation Notes**:
- Core mempool functionality already exists in Gridcoin
- Primarily exposing existing data via RPC
- Start with read-only commands (info, entry, test)
- Add relationship analysis (ancestors/descendants)
- Persistence commands last

**Files to Create/Modify**:
- `src/rpc/mempool.cpp` - NEW FILE (create this)
- `src/rpc/server.cpp` - Register new commands
- May need: `src/txmempool.h/cpp` - Add any missing helper methods

**Reference**:
- Bitcoin: `src/rpc/mempool.cpp`
- Gridcoin: `src/main.cpp` (mempool logic)

---

#### P1.2: PSBT Support (Multisig Workflows)
**Difficulty**: ⭐⭐⭐⭐⭐ (VERY HARD)  
**Estimated Effort**: 3-6 months  
**Impact**: High for multisig users, Foundation, pools

**Commands to Add**:
- `createpsbt` - Create PSBT from inputs/outputs
- `converttopsbt` - Convert raw transaction to PSBT
- `utxoupdatepsbt` - Add UTXO information
- `combinepsbt` - Combine multiple PSBTs
- `finalizepsbt` - Finalize PSBT for broadcast
- `decodepsbt` - Decode PSBT to JSON
- `analyzepsbt` - Analyze PSBT completeness
- `joinpsbts` - Join PSBTs for different inputs
- Wallet integration: PSBT signing commands

**Implementation Approach**:
1. **Phase 1**: Implement PSBT data structures (BIP 174)
   - Create `src/psbt.h/cpp`
   - Port serialization/deserialization
   - ~2-3 weeks

2. **Phase 2**: Basic PSBT operations
   - Create, convert, decode, analyze
   - ~3-4 weeks

3. **Phase 3**: Advanced operations
   - Combine, join, update with UTXO data
   - ~2-3 weeks

4. **Phase 4**: Wallet integration
   - Signing operations
   - Transaction finalization
   - ~3-4 weeks

**Files to Create/Modify**:
- `src/psbt.h/cpp` - NEW FILES (core PSBT implementation)
- `src/rpc/rawtransaction.cpp` - Add PSBT commands
- `src/wallet/wallet.cpp` - PSBT wallet integration
- `src/primitives/transaction.h` - May need extensions

**Dependencies**:
- Good understanding of BIP 174 specification
- May want RPCHelpMan framework first for better docs

**Reference**:
- BIP 174: https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki
- Bitcoin: `src/psbt.h/cpp`, `src/rpc/rawtransaction.cpp`

**Risk Assessment**:
- **High complexity** - PSBT format is complex
- **Testing required** - Needs comprehensive test coverage
- **Breaking changes** - Ensure backward compatibility

---

#### P1.3: Blockchain UTXO Diagnostic Commands (Easy Subset)
**Difficulty**: ⭐⭐ (EASY-MEDIUM)  
**Estimated Effort**: 1-2 weeks  
**Impact**: Medium-High - Useful for debugging

**Commands to Add** (easy ones first):
1. ✅ `getchaintips` - Show all chain tips (EASY, 2-3 days)
2. ✅ `gettxout` - Get UTXO details (EASY, 2-3 days)
3. ✅ `verifychain` - Verify blockchain database (MEDIUM, 3-5 days)

**Implementation Notes**:
- These expose existing Gridcoin functionality
- Mostly read-only queries
- Good first commands for new contributors

**Files to Modify**:
- `src/rpc/blockchain.cpp` - Add these commands
- `src/main.h/cpp` - May need some helper functions

**Reference**:
- Bitcoin: `src/rpc/blockchain.cpp`

---

### 5.4 P2 - Medium Priority (Useful Enhancements)

#### P2.1: Network Diagnostic Commands
**Difficulty**: ⭐⭐ (EASY-MEDIUM)  
**Estimated Effort**: 1-2 weeks  
**Impact**: Medium - Helpful for debugging

**Commands to Add**:
1. ✅ `disconnectnode` - Force disconnect from peer (EASY, 1-2 days)
2. ✅ `setnetworkactive` - Enable/disable networking (EASY, 1-2 days)
3. ✅ `getaddrmaninfo` - Address manager statistics (MEDIUM, 2-3 days)
4. ⚠️ `addpeeraddress` - Manually add peer address (EASY, 1-2 days)
5. ⚠️ `getrawaddrman` - Raw address manager data (MEDIUM, 2-3 days)

**Implementation Notes**:
- Most functionality exists, just needs RPC exposure
- Useful for network debugging
- Low risk

**Files to Modify**:
- `src/rpc/net.cpp` - Add commands
- `src/net.h/cpp` - May need some accessors

**Reference**:
- Bitcoin: `src/rpc/net.cpp`

---

#### P2.2: Node Management Commands
**Difficulty**: ⭐ (EASY)  
**Estimated Effort**: 3-5 days  
**Impact**: Medium - Nice diagnostics

**Commands to Add**:
1. ✅ `uptime` - Server uptime (EASY, 1 day)
2. ✅ `getmemoryinfo` - Memory usage (EASY, 1-2 days)
3. ✅ `getrpcinfo` - RPC call statistics (EASY, 1-2 days)
4. ⚠️ `getindexinfo` - Index status (EASY, 1 day)

**Implementation Notes**:
- Very straightforward information reporting
- Good beginner tasks
- Low complexity

**Files to Create/Modify**:
- `src/rpc/node.cpp` - NEW FILE (or add to misc.cpp)
- Track RPC call stats in `src/rpc/server.cpp`

**Reference**:
- Bitcoin: `src/rpc/node.cpp`

---

#### P2.3: File Reorganization (Wallet RPC)
**Difficulty**: ⭐⭐ (EASY-MEDIUM)  
**Estimated Effort**: 1-2 weeks  
**Impact**: Medium - Better organization

**Goal**:
Create `src/wallet/rpc/` directory and move wallet commands

**Benefits**:
- Clearer code organization
- Easier for contributors to navigate
- Matches Bitcoin structure

**Approach**:
1. Create `src/wallet/rpc/` directory
2. Split wallet commands by category:
   - `addresses.cpp` - Address management
   - `backup.cpp` - Backup/restore
   - `coins.cpp` - UTXO management
   - `encrypt.cpp` - Wallet encryption
   - `spend.cpp` - Sending transactions
   - `transactions.cpp` - Transaction queries
   - `wallet.cpp` - General wallet operations

3. Update build system (CMakeLists.txt, Makefile.am)
4. Test thoroughly

**Dependencies**: Can be done independently or after RPCHelpMan

---

#### P2.4: Blockchain UTXO Analysis (Advanced)
**Difficulty**: ⭐⭐⭐⭐ (HARD)  
**Estimated Effort**: 4-8 weeks  
**Impact**: Medium - Performance improvements

**Commands to Add**:
1. ⚠️ `gettxoutsetinfo` - UTXO set statistics (HARD, 2-3 weeks)
2. ⚠️ `scantxoutset` - Scan UTXO set (VERY HARD, 3-4 weeks)
3. ⚠️ `dumptxoutset` / `loadtxoutset` - UTXO snapshots (VERY HARD, 3-4 weeks)

**Implementation Notes**:
- Requires significant work on UTXO set handling
- Could enable faster sync (assumeutxo)
- Complex but potentially very valuable

**Defer until**: After more critical items

---

#### P2.5: Descriptor Wallet Support
**Difficulty**: ⭐⭐⭐⭐ (HARD)  
**Estimated Effort**: 2-4 months  
**Impact**: Medium - Modern wallet features

**Commands to Add**:
- `getdescriptorinfo` - Parse descriptor
- `deriveaddresses` - Derive addresses
- `createmultisig` - Create multisig from descriptors

**Implementation Notes**:
- Requires implementing descriptor parsing library
- Modern alternative to traditional Bitcoin wallets
- Would enable more flexible wallet setups

**Defer until**: After PSBT support

---

### 5.5 P3 - Low Priority (Nice to Have)

#### P3.1: Transaction Proof Commands
**Difficulty**: ⭐⭐ (EASY-MEDIUM)  
**Estimated Effort**: 1 week  
**Impact**: Low - Niche use case

**Commands**:
- `gettxoutproof` - Create merkle proof
- `verifytxoutproof` - Verify proof

**When**: When there's specific user demand

---

#### P3.2: Transaction Prioritization
**Difficulty**: ⭐ (EASY)  
**Estimated Effort**: 2-3 days  
**Impact**: Low-Medium

**Commands**:
- `prioritisetransaction`
- `getprioritisedtransactions`

**When**: If pool operators request it

---

#### P3.3: Advanced Blockchain Commands
**Difficulty**: ⭐⭐⭐ (MEDIUM-HARD)  
**Estimated Effort**: Variable

**Commands**:
- `pruneblockchain` - Blockchain pruning (HARD, 2-3 weeks)
- `scanblocks` - Block scanning (MEDIUM, 1 week)

**When**: If disk space becomes a major concern (pruning)

---

#### P3.4: Testing Commands (Regtest/Testnet)
**Difficulty**: ⭐ (EASY)  
**Estimated Effort**: 1 week

**Commands**:
- `setmocktime` / `mockscheduler`
- `echo` / `echojson`
- `sendmsgtopeer`
- Mining generation for regtest

**When**: For developer/testing improvements

---

### 5.6 Not Recommended (N/A or Incompatible)

These features are either not applicable to Gridcoin or would require significant architectural changes:

❌ **Fee Estimation Commands** - Gridcoin uses flat fees, not dynamic
❌ **getnetworkhashps** - Not applicable to Proof-of-Stake
❌ **Block template / submitblock** - Gridcoin staking works differently
❌ **External Signer** - Hardware wallet support (very complex, low priority)

---

### 5.7 Recommended Implementation Sequence

**Phase 1: Foundation** (3-6 months)
1. ✅ Adopt RPCHelpMan framework (incremental, start with simple commands)
2. ✅ Add basic mempool commands (getmempoolinfo, getmempoolentry)

**Phase 2: Diagnostics** (1-2 months)
3. ✅ Easy blockchain commands (getchaintips, gettxout, verifychain)
4. ✅ Node management commands (uptime, getmemoryinfo, getrpcinfo)
5. ✅ Basic network diagnostic commands (disconnectnode, setnetworkactive)

**Phase 3: Advanced Mempool** (1 month)
6. ✅ Mempool analysis (ancestors, descendants)
7. ✅ Mempool persistence (save/import)
8. ✅ Transaction testing (testmempoolaccept)

**Phase 4: Code Organization** (1-2 months)
9. ✅ Reorganize wallet RPC into src/wallet/rpc/
10. ✅ Create dedicated mempool.cpp, node.cpp files

**Phase 5: Major Features** (3-6 months each)
11. ⚠️ PSBT support (if multisig users need it)
12. ⚠️ Descriptor wallet (if modern wallet features desired)

**Phase 6: Polish** (ongoing)
13. ⚠️ Additional diagnostic commands as needed
14. ⚠️ Performance optimizations (UTXO set operations)

---

### 5.8 Quick Wins (Low-Hanging Fruit)

These commands can be added quickly by new contributors:

**Week 1-2 Tasks** (Good for first-time contributors):
- `uptime` (1 day)
- `getchaintips` (2-3 days)
- `disconnectnode` (1-2 days)
- `setnetworkactive` (1-2 days)

**Week 3-4 Tasks** (Slightly more complex):
- `getmempoolinfo` (2-3 days)
- `getmempoolentry` (2-3 days)
- `gettxout` (2-3 days)
- `getmemoryinfo` (2-3 days)

---

### 5.9 Effort vs Impact Matrix

```
High Impact    │ P1.1 Mempool      │ P0.1 RPCHelpMan
               │ P1.3 UTXO Easy    │ P1.2 PSBT
               ├───────────────────┼─────────────────
               │ P2.1 Net Diag     │ P2.4 UTXO Hard
               │ P2.2 Node Mgmt    │ P2.5 Descriptors
Low Impact     │ P3.1 Proofs       │
               └───────────────────┴─────────────────
                 Easy/Medium         Hard/Very Hard
```

**Strategy**: Focus on top-left (high impact, easier) first, then move right.

---

## Section 6: Conclusion and Recommendations

### 6.1 Summary

**Current State**:
- Gridcoin has ~140 RPC commands covering core functionality
- Strong support for Gridcoin-specific features (staking, research, voting)
- Gaps primarily in Bitcoin's general-purpose diagnostic and advanced features

**Major Gaps**:
1. **Framework**: No RPCHelpMan self-documenting system
2. **Mempool**: Minimal diagnostic commands (only basic listing)
3. **PSBT**: No support for modern multisig workflows
4. **Organization**: Could benefit from better file structure

**Strengths**:
- Comprehensive Gridcoin-specific functionality
- Core wallet operations well-implemented
- Network and blockchain basics covered

### 6.2 Top Recommendations

#### Recommendation 1: Start with RPCHelpMan Framework (P0)
**Why**: Makes all future RPC work easier and better documented  
**Approach**: Incremental conversion, start with simple utility commands  
**Timeline**: Begin immediately, complete over 3-6 months

#### Recommendation 2: Add Mempool Diagnostics (P1)
**Why**: Important for debugging, relatively quick wins  
**Approach**: Start with getmempoolinfo and getmempoolentry  
**Timeline**: 2-4 weeks for basic commands

#### Recommendation 3: Easy Blockchain/Network Commands (P1/P2)
**Why**: Good for new contributors, useful functionality  
**Approach**: Add getchaintips, gettxout, disconnectnode, etc.  
**Timeline**: 1-2 weeks per command

#### Recommendation 4: Evaluate PSBT Need (P1)
**Why**: Important for multisig users (Foundation, pools)  
**Approach**: Survey community need first, then plan implementation  
**Timeline**: 3-6 months if needed

#### Recommendation 5: Code Organization (P2)
**Why**: Better long-term maintainability  
**Approach**: Create src/wallet/rpc/ directory, reorganize files  
**Timeline**: 1-2 weeks

### 6.3 Community Engagement

**Before starting major work**:
1. Create GitHub issues for each major feature
2. Get community feedback on priorities
3. Identify willing contributors
4. Document implementation plans

**Good first issues** for new contributors:
- Node management commands (uptime, getmemoryinfo)
- Simple blockchain queries (getchaintips, gettxout)
- Network diagnostics (disconnectnode, setnetworkactive)

### 6.4 Long-Term Vision

**Year 1**: Foundation and quick wins
- RPCHelpMan framework adoption
- Mempool diagnostic commands
- Easy blockchain/network commands
- Code reorganization

**Year 2**: Major features (if community wants them)
- PSBT support
- Descriptor wallet features
- Advanced UTXO set operations

**Ongoing**: Maintain parity with Bitcoin where applicable while preserving Gridcoin's unique features.

---

## Appendices

### Appendix A: Command Difficulty Legend

- ⭐ **EASY** (1-3 days): Simple information reporting, existing functionality
- ⭐⭐ **MEDIUM** (3-7 days): Some new logic, moderate complexity
- ⭐⭐⭐ **HARD** (1-3 weeks): Significant new code, complex logic
- ⭐⭐⭐⭐ **VERY HARD** (1-3 months): Major feature, extensive work
- ⭐⭐⭐⭐⭐ **EXTREMELY HARD** (3-6 months): Architectural changes, high complexity

### Appendix B: Bitcoin RPC References

**Primary Source**:
- Bitcoin GitHub: https://github.com/bitcoin/bitcoin
- RPC Documentation: https://developer.bitcoin.org/reference/rpc/

**Key Files to Reference**:
- `src/rpc/server.h/cpp` - Command registration
- `src/rpc/util.h/cpp` - RPCHelpMan framework
- `src/rpc/*.cpp` - Individual command implementations

### Appendix C: Related Gridcoin Issues

- **Issue #1142**: Original request to port RPC functionality from Bitcoin
- **Issue #1705** & **PR #1995**: Created src/rpc/ directory structure (v5.2.0.0)

### Appendix D: Contributor Resources

**Getting Started**:
1. Read Gridcoin CONTRIBUTING.md
2. Set up development environment
3. Start with P3 "quick wins" commands
4. Join Discord developer channel

**Documentation**:
- `.clinerules/` directory for codebase context
- `doc/` directory for build instructions
- Bitcoin Core documentation for RPC patterns

---

**End of Report**

---

**Document Version**: 1.0  
**Last Updated**: December 27, 2025  
**Prepared by**: RPC Porting Analysis  
**For**: Gridcoin Development Community

For questions or to contribute: https://github.com/gridcoin-community/Gridcoin-Research

**Comment by jamescowens** [comment]
This is an extremely useful analysis and implementation plan. We need to have further discussion on how to control the regression risk, especially around RPCHelpMan. Perhaps we can do this in substages. We need to implement PBST as multisig transactions are done by passing the raw tx data around until it is completed.

**Comment by jamescowens** [comment]
## Re-scoping #1142 after a detailed RPC-divergence analysis

This issue (2018) had three goals: **(1)** port Bitcoin's RPC engine into `rpc/`; **(2)** isolate Gridcoin-specific calls into `src/gridcoin/rpc/` (Bitcoin naming); **(3)** inject them via a `RegisterGridcoinRPCCommands` called from core registration. We audited the current tree against all three. Summary up front:

- **Goal 1 — DONE.** The modern Bitcoin RPC engine is in place (`CRPCCommand` table, `RPCHelpMan` everywhere after #2922, Bitcoin file-naming under `src/rpc/`, `client.cpp` conversion table). The "copy it and see if it cascades" feasibility question is long resolved.
- **Goals 2 & 3 — re-scoped / largely declined** for the reasons below. They optimize *provenance* (did this originate in Bitcoin?) when the thing we actually need is *fidelity* (is this still upstream-faithful — can I safely backport into it?). Those are not the same question, and a folder/registration split answers the wrong one.

### What the analysis found

There are ~185 RPC commands. **56 share a name with upstream Bitcoin; 129 are Gridcoin-only** (no upstream namesake at all). The 56 "looks-like-Bitcoin" set is where the risk lives, so we measured it on two axes:

**Input axis (argument surface).** Comparing `rpc/client.cpp` arg-conversion fingerprints, Gridcoin vs current upstream:
- Gridcoin carries **150** arg-conversions vs upstream's **336**, and is *systematically the smaller/older* surface. 27 of the 56 have visibly diverged arg index sets; e.g. `sendtoaddress` GRC{1} vs CORE{0–10}, `sendmany` {1,2} vs {0–9}, `getbalance` missing `avoid_reuse`, `gettransaction` missing `verbose`.
- This divergence is dominated by **staleness** — we froze at an old, smaller Bitcoin surface and never renovated. Critically, it is **visible in the signature** (and machine-comparable post-#2922).

**Output / behavior axis (the RPC body).** This is the axis a folder split cannot see:
- `src/rpc/blockchain.cpp` emits **80** Gridcoin-specific result fields (`claim`, `cpid`, `magnitude`, `superblock`, `beacons`, `mint`, `IsSuperBlock`, …); `mining.cpp` emits `accrual*`, `BoincRewardPending`, `netstakeweight`, `current_magnitude`, …; `getmininginfo` is aliased to `getstakinginfo` (a PoW name reused for a PoS implementation).
- **Poster child: `getblock`.** Its *argument surface matches upstream* (so it isn't in the diverged-arg set) — yet its body returns `claim`/`cpid`/`IsSuperBlock`/`BoincHash`. That is **required** divergence (intrinsic to PoS + research-reward + superblock consensus), it is **permanent** (no renovation removes it), and it is **invisible at the surface**. `getblockchaininfo`, `getdifficulty` (PoS), `gettransaction` (contract data), `validateaddress`, etc. are the same shape. The accounts API (`getaccount`/`move`/`sendfrom`/…) and `getinfo` were *removed* upstream entirely and we retain them.

### Why the folder/registration separation is artificial, misleading, and futile

- **Artificial** — it sorts by provenance, but provenance doesn't track fidelity. Moving the obviously-Gridcoin RPCs out of `blockchain.cpp` does **not** make `getblock` backport-clean; it still emits superblock data.
- **Misleading** — a `src/rpc/ = Bitcoin` taxonomy *connotes* upstream-faithful. The most dangerous divergence (required result customization) is exactly the kind invisible at the arg surface, so the folder label would actively mislead someone trying to sync upstream. For the required-diverged RPCs that's a **standing lie**, not a transient one.
- **Futile** — the only RPCs you could safely backport into are the genuinely-faithful ones, and the analysis suggests that bucket is *small*: most of the 56 shared-name RPCs are either stale-and-modified or required-diverged. So the isolation buys a comforting label over an almost-empty set of "safe" files.

### Recommended replacement: a three-bucket RPC heritage ledger

Instead of a folder taxonomy, track each RPC's **heritage** explicitly, because that is what a backporter actually needs to know:

1. **Pure upstream backport** — faithful to current upstream; an upstream change can be taken with normal care. (Expected to be a *small* set.)
2. **Mixed heritage** — body partly upstream-derived but carries significant Gridcoin differences (PoS/research/contract output, retained-old args, removed-upstream semantics). An upstream change requires **careful manual porting** — this is the dangerous "looks portable, isn't" bucket, and the whole point of the ledger is to flag it.
3. **Pure Gridcoin** — no upstream analogue (the ~129). Upstream changes are N/A.

**Form (in code vs doc):** recommend a single checked-in canonical table (e.g. `doc/rpc-heritage.md`) for one-place auditability, **plus** a lightweight one-line in-code marker per RPC (e.g. `// RPC-HERITAGE: mixed — diverges: emits superblock/claim`) and a small lint that cross-checks the markers against the doc so the two cannot silently drift. Doc gives the reviewable overview; the in-code tag keeps the truth next to the code; the lint prevents rot.

### Disposition of the original goals
- **Goal 1:** done — close out.
- **Goal 3 (`RegisterGridcoinRPCCommands`):** *limited value on its own.* It only addresses **bucket 3** (the pure-Gridcoin commands), adds registration structure + ongoing maintenance, and does **nothing** for the actual pain (knowing which *inherited* RPCs in buckets 1/2 are safe to touch). Its one forward-looking justification is a cleaner registration seam for the multiprocess split (#2937); absent a concrete need there, **deprioritize**.
- **Goal 2 (hoist Gridcoin RPCs into `src/gridcoin/rpc/`):** *limited value; opportunistic only.* If a bucket-3 RPC is being touched anyway, moving it can be fine — but never as a big-bang reorg, and never sold as making `src/rpc/` upstream-faithful.

**Net:** keep #1142 open but re-scoped to "build the three-bucket RPC heritage ledger (doc + in-code markers + drift lint)"; treat the original folder-isolation/registration-split as low-priority/opportunistic. The ledger is the artifact that actually serves backport hygiene; the folder never could.

**Comment by jamescowens** [comment]
Closing this long-standing (2018) issue. **Goal 1 (port Bitcoin's modern RPC engine) is complete.** Goals 2 & 3 (folder isolation + `RegisterGridcoinRPCCommands`) are superseded — per the analysis above, provenance-based separation is artificial, misleading, and futile (it can't encode the required, surface-invisible, permanent divergence in RPC bodies). The genuinely-useful successor — a **3-bucket RPC heritage ledger** — is tracked in #3069. Closing as the core deliverable landed and the remainder is consciously redirected, not abandoned.
